Dockerfile for nvitop-exporter

gianfranco-s opened the pull request:
<!-- Provide a descriptive summary of the changes in the title above -->

#### Issue Type
- Improvement/feature implementation

#### Runtime Environment

_Not really necessary when using Docker._

- Operating system and version: Pop!_OS 22.04 LTS
- Terminal emulator and version: GNOME Terminal 3.44.0 
- Python version: 3.11.7
- NVML version (driver version): 560.35.03
- `nvitop` version or commit: main@254716f1aeaa656401e3978112ad64be6be007f4
- `python-ml-py` version: -
- Locale: en_US.UTF-8

#### Description
* Added Dockerfile to build image for nvitop-exporter
* Added instructions to start exporter
* Updated nvitop Dockerfile due to warning about ENV label

#### Motivation and Context
Changes required to start the exporter in a docker container, without interactive installs.


#### Testing
* Metrics are exposed correctly in http://localhost:5050. It can be consumed normally by Grafana/PRometheus

No other areas are affected.

gianfranco-s commented:
I get that what you’re aiming for is a Grafana dashboard that works out of the box.

Given that Grafana and nvitop aren't always deployed on the same machine, I wouldn't go for a single image bundling all services (nvitop/nvitop-exporter, Prometheus, Grafana). Instead, I'd use a specific dasboard-deploy directory, with the required services necessary to consume the data from the exporter.

Something like
 .
├── .github
├── docs
├── nvitop-exporter
├── **nvitop-grafana**
├── nvitop
├── ...
├── requirements.txt
└── setup.py

If it's ok with you, I can propose that solution on a different PR; the nvitop-grafana feature is related to the current PR, but not really necessary for it.

XuehaiPan commented:
> I wouldn't go for a single image bundling all services

I mean we need a single image for `nvitop-exporter` with an additional `docker-compose.yml` example.

gianfranco-s commented:
This is what I mean (the PR is a WIP; still trying to set up the provisioned dashboard):
https://github.com/gianfranco-s/nvitop/pull/2

XuehaiPan commented on nvitop-exporter/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM nvitop:latest
```
Just build from an NVIDIA driver image. `pipx` will automatically setup the Python dependencies.

gianfranco-s replied:
Nice!

gianfranco-s commented:
**Updates**
* Moved dashboard.json to the new nvitop-grafana/ directory
* Added docker-compose.yml and provisioned Prometheus scraper

@XuehaiPan please review again.
With the current solution, all three services (Grafana, Prometheus and nvito-exporter) start with one command; no need for additional building steps.

vashian commented:
PING @XuehaiPan

vashian commented:
> **Updates**
> 
> * Moved dashboard.json to the new nvitop-grafana/ directory
> * Added docker-compose.yml and provisioned Prometheus scraper
> 
> @XuehaiPan please review again. With the current solution, all three services (Grafana, Prometheus and nvito-exporter) start with one command; no need for additional building steps.

I’m looking to deploy this exporter as a DaemonSet in Kubernetes. Do you have any experience with this?

XuehaiPan commented:
Hi @gianfranco-s, sorry for the late review. I have updated the PR, please check if your are happy with it.

Now you can start the dashboard service with a single command:

```bash
docker compose --project-directory=nvitop-exporter/grafana up --build --detach
```

You will not need to import the dashboard file manually.
